File: javascript/slotmachine.js
// deposit money
// determine lines
// collect bet amount
// play the game
// check for wins
// return output and/or winning
// check if repeatable and ask for new gameplay

const prompt = require("prompt-sync")();

const rows = 3;
const cols = 3;

const symbols_count = {
    "$":2,
    "@":4,
    "%":6,
    "#":8
}

const symbols_values = {
    "$":5,
    "@":4,
    "%":3,
    "#":2
}

//deposit money
const getDeposit = () => {
    while (true){
        const stramount = prompt("Enter a deposit amount: $");
        const amount = parseFloat(stramount);

        if (isNaN(amount) || amount <= 0){
            console.log("Invalid deposit amount. Try again");
        }
        else{return amount;}
    }
}   

const getLines = () => {
    while (true){
        const strlines = prompt("Enter the number of lines to place a bet on (1-3): ");
        const lines = parseFloat(strlines);

        if (isNaN(lines) || lines <= 0 || lines > 3){
            console.log("Invalid number of lines. Try again");
        }
        else{return lines;}
    }
}   

const getBet = (balance,lines) => {
    while (true){
        const strbet = prompt("How much would you like to bet per line: ");
        const bet = parseFloat(strbet);

        if (isNaN(bet) || bet <= 0 || bet > balance/lines){
            console.log("Invalid bet amount. Try again");
        }
        else{return bet;}
    }
}

const spin = () => {
    const symbols = [];
    for(const [symbol, count] of Object.entries(symbols_count)){
        for (let i = 0; i < count; i++){
            symbols.push(symbol);
        }
    }
    const reels = [];
    for (let a = 0; a < cols; a++){
        reels.push([]);
        const reelSymbols = [...symbols];
        for (let b=0; b < rows; b++){
            const randIndex = Math.floor(Math.random() * reelSymbols.length);
            const selectedsymbol = reelSymbols[randIndex]
            reels[a].push(selectedsymbol);
            reelSymbols.splice(randIndex, 1);
        }
    }
    return reels;
};

const transpose = (reels) => {
    const row = [];
    
    for (let i = 0; i < rows; i++){
        row.push([]);
        for (let j = 0; j < cols; j++){
            row[i].push(reels[j][i])
        }
    }
    return row
}

const result = (transrows) => {
    console.log();
    console.log("Here is your play result ");
    for (const row of transrows){
        let rowString = '';
        for (const[i, symbol] of row.entries()){
            rowString += symbol
            if (i != row.length - 1){
                rowString += " | "
            }
        }
        console.log(rowString);
    }
    
}

const getWinnings = (transrows, bets, lines) => {
    let winnings = 0;

    for (let row = 0; row < lines; row++){
        const symbols = transrows[row];
        let similar = true;

        for (const symbol of symbols){
            if (symbol != symbols[0]){
                similar = false;
                break;
            }
        }

        if (similar){
            winnings += (bets * symbols_values[symbols[0]])
        }
    }
    return winnings;
}

const game = () => {
    let balance = getDeposit();

    while(true){
        
        const lines = getLines();
        const bets = getBet(balance, lines);
        balance -= bets * lines
        const reels = spin();
        const transrows = transpose(reels);
        //console.log(reels);
        //console.log(transrows);
        result(transrows);
        const winnings = getWinnings(transrows, bets, lines);
        balance += winnings;
        console.log();
        console.log("You won $" + winnings.toString() + " and spent $" + (bets * lines));
        console.log("You have a balance of $" + balance);

        if (balance <= 0){
            console.log("You ran out of money");
            break;
        }

        const playagain = prompt("Enter any key to continue or 'n' to stop: ");
        if (playagain == "n"){break;}

    }
}

game();

File: C/slotmachine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>
#include <unistd.h>
#include "mwendwa.h"

#define size 100
char symbols[4] = {'#','@','$','!'};
int symbolsValue[] = {2,4,6,8};
char* slotResult;
char* analResult;



// deposit money
int getDeposit(){
    int amount = 0;
    while (1){
        printf("Please enter your deposit amount: ");
        if (scanf("%d",&amount) != 1 ||amount < 1 ){
            printf("Please enter a valid deposit amount.\n");
            while (getchar() != '\n') continue;
        }
        else{break;}
    }
    return amount;
}

// determine lines
int getLines(){
    int lines = 0;
    while (1){
        printf("Please enter the number of lines to bet on (1-3): ");
        if (scanf("%d",&lines) != 1 ||lines < 1 || lines > 3 ){
            printf("Please enter a valid number of lines.\n");
            while (getchar() != '\n') continue;
        }
        else{break;}
    }
    return lines;
}

// collect bet amount
int getBet(int amount,int lines){
    int bet = 0;
    int maxBet = amount/lines;
    while (1){
        printf("Please enter the amount to bet per line: ");
        if (scanf("%d",&bet) != 1 || bet < 1 ){
            printf("Please enter a valid deposit bet.\n");
            while (getchar() != '\n') continue;
        }
        else if (bet > (maxBet)){
            printf("Your bet amount cannot be greater than %d.\n", maxBet);
            continue;
        }
        else{break;}
    }
    return bet;
}

// play the game
char* slot(char symbols[], char* slotResult){
    srand(time(NULL));
    slotResult = malloc(size * sizeof(char));

    
    for(int i=0; i<3; i++){
        for (int j=0; j<3; j++){
            int index = rand() % 4;
            append(slotResult, '|');
            append(slotResult, ' ');
            append(slotResult, symbols[index]);
            //I could make the header file do these lines but a good programmer is a lazy programmer :)
            append(slotResult, ' ');
            append(slotResult, '|');
            append(slotResult, ' ');
        }
        append(slotResult, '\n');
    }
return (slotResult);
}

// analyze the result
char* analSlot(char symbols[], char* analResult){
    srand(time(NULL));
    analResult = malloc(size * sizeof(char));

    
    for(int i=0; i<3; i++){
        for (int j=0; j<3; j++){
            int index = rand() % 4;
            append(analResult, symbols[index]);
        }
    }
return (analResult);
}
// check for wins
int getWinning(char* analResult,int bet, char symbols[], int symbolsValue[]){
    int winning = 0;
    char win = analyze(analResult);
    //printf("%c, %s\n",win, analResult); //for testing purposes, remove this

    for (int i=0; i<4; i++){
        if (win == symbols[i]){
            winning = bet * symbolsValue[i];
        }
    }
    return winning;

}

// return output and/or winning
// check if repeatable and ask for new gameplay

int main(void)
{
    int amount = getDeposit();
    while (1){
        
        int lines = getLines();
        int bet = getBet(amount, lines);
        int betAmount = bet * lines;

        printf("You bet %d on %d line(s).\nYour total bet amount is %d shillings\n", bet, lines, betAmount);
        char* result = slot(symbols, slotResult);
        char* analREsult = analSlot(symbols, analResult);

        int win = getWinning(analREsult, betAmount, symbols, symbolsValue);
        
        sleep(2);
        printf("\n%s\n",result);
        //printf("%d - %d - %d\n",amount, lines, bet); //for testing purposes, remove this
        //printf("%s\n", analResult); //for testing purposes, remove this
        //printf("%d",win); //for testing purposes, remove this

        if (win == 0){
            printf("You lost %d shillings\n", betAmount);
            amount = amount - betAmount;
        }
        else{
            printf("You won %d shillings!\n", win);
            amount = amount + win;
        }
        printf("Your new balance is %d shillings\n\n", amount);
        if (amount == 0){printf("You ran out of money!\nGerarra here brokie \n");break;}

        char playAgain;
        printf("Do you want to play again? (y/n): ");
        scanf(" %c", &playAgain);
        printf("\n");
        if (tolower(playAgain) == 'y'){
            continue;
        }
        else{
            printf("Thank you for playing.\n");
            break;
        }
    }
}